docs/architecture/adr-018-network-upgrades.md

```diff
@@ -29,40 +30,48 @@ One of the main difficulties of social upgrades when using tendermint consensus
+- Given uncertainty in scheduling, the system must be able to handle changes to the upgrade height that most commonly would come in the form of delays. Embedding the upgrade schedule in the binary is convenient for node operators and avoids the possibility for user errors. However, binaries are static. If the community wished to push back the upgrade by two weeks there is the possibility that some nodes would not rerun the new binary thus we'd get a split between nodes running the old schedule and nodes running the new schedule. To overcome this, proposers will only propose a version change in the first round of each height, thus allowing transactions to still be committed even under circumstances where there is no consensus on upgrading. Secondly, we define a range in which nodes will attempt to upgrade the app version and failing this will continue to run the current version. Lastly, the binary will have the ability to manually specify the app version height mapping and overide the built-in values either through a flag or in the `app.toml` config. This is expected to be used in testing and in emergency situations only. Another example to keep in mind is if a quorum outright rejects an upgrade. If some of the validators are for the change they should have some way to continue participating in the network. Therefore we employ a range that nodes will attempt to upgrade and afterwards will continue on normally with the new binary however running the older version.
```

Comment:
There shouldn't be overlap with ranges (currently the node errors on start up if there is).

The reason for an end range is because at some point the node should give up trying to propose an upgrade (as no other node wants to do it and it's costing us unnecessary rounds). An example of why this is important is say we release a version with a bug in it. The initial heights for the upgrade are [100, 200]. When we notice the bugs we tell node operators to cancel the upgrade. So long as 1/3+ clear the upgrade schedule, the upgrade proposal will always be rejected. After the end height, the nodes who didn't clear it will stop proposing the upgrade and continue to produce normal v1 blocks. If we didn't have the end height we would continually see multi-round blocks which are slower.

On this topic:
- It takes 1/3+ to deny an upgrade
- It takes 2/3+ to perform an upgrade

(maybe I should mention that here somewhere)

Reply:
Ahh if ranges are non-overlapping then I think it's easier to reason about. I assume there can be gaps in the range. Ex.

```
v2 = [100, 200]
v3 = [300, 400]
```

Reply:
Yes that's fine